### PR TITLE
Use callbacks during event scheduling

### PR DESCRIPTION
**Pull request by rlazo**

### CODE REVIEW

File: transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/TransportImpl.java [hunk 1]
```diff
@@ -36,8 +37,8 @@
   }
 
   @Override
-  public void send(Event<T> event) {
```
Comment (vkryachko) [inline transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/TransportImpl.java]: We don't want to remove this method to preserve backwards compatibility. pls add a new method, maybe `schedule()` or `sendAsync()`

Comment (rlazo) [reply transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/TransportImpl.java]: PTAL

File: transport/transport-api/transport-api.gradle [hunk 2]
```diff
@@ -30,6 +30,7 @@ android {
 }
 
 dependencies {
+    compileOnly 'com.google.android.gms:play-services-tasks:17.0.0'
```
Comment (vkryachko) [inline transport/transport-api/transport-api.gradle]: pls remove this dep

File: transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/ImmediateScheduler.java [hunk 3]
```diff
@@ -43,17 +44,26 @@ public ImmediateScheduler(Executor executor, BackendRegistry backendRegistry) {
   }
 
   @Override
-  public void schedule(TransportContext transportContext, EventInternal event) {
```
Comment (vkryachko) [inline transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/ImmediateScheduler.java]: not completely related, but can you pls remove this class from the `main` source-set and put it into test, it's not used in the sdk, only in tests

File: transport/transport-runtime/src/test/java/com/google/android/datatransport/runtime/TransportRuntimeTest.java [hunk 4]
```diff
@@ -47,6 +55,26 @@
...
+  private final EventStore mockEventStore = mock(EventStore.class);
+  private static final SynchronizationGuard guard =
+      new SynchronizationGuard() {
+        @Override
+        public <T> T runCriticalSection(CriticalSection<T> criticalSection) {
+          return criticalSection.execute();
+        }
+      };
+
+  private class StatefulTransportScheduleCallback implements TransportScheduleCallback {
```
Comment (vkryachko) [inline transport/transport-runtime/src/test/java/com/google/android/datatransport/runtime/TransportRuntimeTest.java]:
```suggestion
  private static class StatefulTransportScheduleCallback implements TransportScheduleCallback {
```

File: transport/transport-api/src/main/java/com/google/android/datatransport/Transport.java [hunk 5]
```diff
@@ -23,7 +25,22 @@
...
+   * events are actually send.
+   *
    * @param event The event with the payload that needs to be sent.
    */
   void send(Event<T> event);
+
+  /**
+   * Schedules sending the event of type T.
+   *
+   * <p>The returned {@code Task} tracks the process of persisting the event to permanent storage to
```
[the remark this answers is not in the document]
Comment (rlazo) [reply transport/transport-api/src/main/java/com/google/android/datatransport/Transport.java]: Done

File: transport/transport-api/transport-api.gradle (hunk 2 above)
Comment (rlazo) [reply transport/transport-api/transport-api.gradle]: Done.

File: transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/ImmediateScheduler.java (hunk 3 above)
Comment (rlazo) [reply transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/ImmediateScheduler.java]: Done.

File: transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/DefaultScheduler.java [hunk 6]
```diff
@@ -60,24 +61,33 @@
...
+            EventInternal decoratedEvent = transportBackend.decorate(event);
+            guard.runCriticalSection(
+                () -> {
+                  eventStore.persist(transportContext, decoratedEvent);
+                  workScheduler.schedule(transportContext, 1);
+                  return null;
+                });
+            callback.onSchedule(null);
+          } catch (Exception e) {
+            callback.onSchedule(e);
```
Comment (vkryachko) [inline transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/DefaultScheduler.java]: Before this change the exception would have been logged, can you log it here to preserve this behavior?

Comment (rlazo) [reply transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/DefaultScheduler.java]: Done.

**Pull request merged**